command/operator_diagnose.go

```diff
@@ -145,87 +157,68 @@ func (c *OperatorDiagnoseCommand) RunWithParsedFlags() int {
 
-	// Make sure we close all listeners from this point on
-	listenerCloseFunc := func() {
-		for _, ln := range lns {
-			ln.Listener.Close()
+		if status != 0 {
+			return errMsg
 		}
-	}
 
-	defer c.cleanupGuard.Do(listenerCloseFunc)
-
-	sanitizedListeners := make([]listenerutil.Listener, 0, len(config.Listeners))
-	for _, ln := range lns {
-		if ln.Config.TLSDisable {
-			c.UI.Warn("WARNING! TLS is disabled in a Listener config stanza.")
-			continue
-		}
-		if ln.Config.TLSDisableClientCerts {
-			c.UI.Warn("WARNING! TLS for a listener is turned on without requiring client certs.")
+		// Make sure we close all listeners from this point on
+		listenerCloseFunc := func() {
+			for _, ln := range lns {
+				ln.Listener.Close()
+			}
 		}
 
-		// Check ciphersuite and load ca/cert/key files
-		// TODO: TLSConfig returns a reloadFunc and a TLSConfig. We can use this to
-		// perform an active probe.
-		_, _, err := listenerutil.TLSConfig(ln.Config, make(map[string]string), c.UI)
-		if err != nil {
-			c.UI.Output("Error creating TLS Configuration out of config file: ")
-			c.UI.Output(err.Error())
-			return 1
-		}
-
-		sanitizedListeners = append(sanitizedListeners, listenerutil.Listener{
-			Listener: ln.Listener,
-			Config:   ln.Config,
-		})
-	}
-	err = diagnose.ListenerChecks(sanitizedListeners)
-	if err != nil {
-		c.UI.Output("Diagnose caught configuration errors: ")
-		c.UI.Output(err.Error())
-		return 1
-	}
+		defer c.cleanupGuard.Do(listenerCloseFunc)
 
-	// Errors in these items could stop Vault from starting but are not yet covered:
-	// TODO: logging configuration
-	// TODO: SetupTelemetry
-	// TODO: check for storage backend
-	c.UI.Output(same_line + status_ok + phase)
-
-	phase = "Access storage"
-	c.UI.Output(status_unknown + phase)
-	_, err = server.setupStorage(config)
-	if err != nil {
-		c.UI.Output(same_line + status_failed + phase)
-		c.UI.Output(err.Error())
-		return 1
-	}
-	c.UI.Output(same_line + status_ok + phase)
+		sanitizedListeners := make([]listenerutil.Listener, 0, len(config.Listeners))
+		for _, ln := range lns {
+			if ln.Config.TLSDisable {
+				diagnose.Warn(ctx, "TLS is disabled in a Listener config stanza.")
+				continue
+			}
+			if ln.Config.TLSDisableClientCerts {
+				diagnose.Warn(ctx, "TLS for a listener is turned on without requiring client certs.")
+			}
+
+			// Check ciphersuite and load ca/cert/key files
+			// TODO: TLSConfig returns a reloadFunc and a TLSConfig. We can use this to
+			// perform an active probe.
+			_, _, err := listenerutil.TLSConfig(ln.Config, make(map[string]string), c.UI)
+			if err != nil {
+				return err
+			}
+
+			sanitizedListeners = append(sanitizedListeners, listenerutil.Listener{
+				Listener: ln.Listener,
+				Config:   ln.Config,
+			})
+		}
+		return diagnose.ListenerChecks(sanitizedListeners)
+	})
 
-	return 0
+	return diagnose.Test(ctx, "storage", func(ctx context.Context) error {
+		_, err = server.setupStorage(config)
+		return err
+	})
+	return nil
```

Comment:
hmm, it looks like you're returning right after another return?

Reply:
this github preview doesn't show enough code to show what i'm pointing out

Reply:
Can you take another look?  It's tricky because there is an anonymous function there as an argument to diagnose.Test, so you're probably seeing that functions return followed by the other one.

Reply:
<img width="701" alt="Screen Shot 2021-04-29 at 9 41 55 AM" src="https://user-images.githubusercontent.com/19847032/116578946-23f6b700-a8cf-11eb-838b-da77ce5baae5.png">


yeah, this is what i'm seeing. I'm not thinking about the `return err` bit inside the `func(ctx context.Context) error`, i'm thinking about the `return diagnose.Test...` on line 219

thoughts?

Reply:
basically i think the return nil is a noop - and i'm guessing the compiler doesn't complain about that if the tests are running?